### PR TITLE
Finish find meeting algorithm and add test cases

### PR DESCRIPTION
**Pull request by yjzhang111**
Complete algorithm in FindMeetingQuery.java and add more edge cases to FindMeetingQueryTest.java

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -270,5 +272,149 @@ public void notEnoughRoom() {
...
+  public void optionalAttendeeNotConsidered() {
+    // Have each person have different events. We should see three options because each person has
+    // split the restricted times.
+    // Optional attendee has all day event. We should still return the three options
+    //
+    // Events  :       |--A--|     |--B--|
+    // Day     : |-----------------------------|
+    // Options : |--1--|     |--2--|     |--3--|
+
+    Collection<Event> events = Arrays.asList(
```
Comment (albert-g3) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Since it's a list anyway I recommend instantiating to List here to allow for easier looping over it (you wouldn't need to use Iterator)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -270,5 +272,149 @@ public void notEnoughRoom() {
...
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
```
Comment (albert-g3) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: I know you didn't name FindMeetingQuery, it's not a good name, it goes against the convention.  Class names should be nouns, so something like MeetingFinder.  Then  query.query can be `meetingFinder.query` which is easier to read and understand what's suppose to happen here.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 3]
```diff
@@ -270,5 +272,149 @@ public void notEnoughRoom() {
...
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
```
Comment (albert-g3) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Consider static import as such, https://docs.oracle.com/javase/7/docs/technotes/guides/language/static-import.html
Google recommends import static methods when it doesn't decrease readability

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,100 @@
...
+/*  Find time slot(s) that all the required attendees are availble
+    Include optional attendees if at least one time slot can be found
+    Otherwise return the time slots that fit just the mandatory attendees.
+*/
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Initialize a time slot (the whole day)
+    List<TimeRange> timeSlots = new ArrayList<>();
+    timeSlots.add(TimeRange.fromStartDuration(0, 24 * 60));
```
Comment (albert-g3) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Can you extract 24 * 60 to a static variable? Generally avoid numbers in code unless it's really simple or provide a comment.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,100 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Iterator;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
 
+
+/*  Find time slot(s) that all the required attendees are availble
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: typo: available.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,100 @@
...
+    List<TimeRange> timeSlots = new ArrayList<>();
+    timeSlots.add(TimeRange.fromStartDuration(0, 24 * 60));
+
+    // Find time slots for required attendees
+    timeSlots = queryForEvent(events, timeSlots, request.getAttendees());
+    timeSlots = deleteUnqualifiedQuery(timeSlots, request.getDuration());
+
+    // Save the result in timeSlotsForRequired
+    // in case no time slot can be found when optional attendees are included
+    List<TimeRange> timeSlotsForRequired = new ArrayList<TimeRange>(timeSlots);
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:46]: Is it necessary to convert timeSlots to an ArrayList?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,100 @@
...
+
+    // Check for the availability of optional attendees
+    timeSlots = queryForEvent(events, timeSlots, request.getOptionalAttendees());
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: Does "timeSlots" include slots for required attendees, or only optional attendees?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -15,9 +15,100 @@
...
 import java.util.Collection;
+import java.util.Iterator;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
 
+
+/*  Find time slot(s) that all the required attendees are availble
+    Include optional attendees if at least one time slot can be found
+    Otherwise return the time slots that fit just the mandatory attendees.
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: the documentation uses 'mandatory' and 'required' interchangeably. Not a big deal, but it's a good idea to stay consistent with terminology.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -15,9 +15,100 @@
...
+*/
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: 1) Generally best practice to document public methods, classes, interfaces, etc.
2) What does the method "query" do if no meeting slots are found? Should we document this?
3) Another case: what happens if required attendees have no available time slots, but optional attendees do?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -15,9 +15,100 @@
...
+    timeSlots = queryForEvent(events, timeSlots, request.getOptionalAttendees());
+    timeSlots = deleteUnqualifiedQuery(timeSlots, request.getDuration());
+
+    if (timeSlots.isEmpty() && !request.getAttendees().isEmpty()) {
+      return timeSlotsForRequired;
+    }
+    return timeSlots;
+  }
+
+  // Iterate through all event to check
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: event --> events

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -15,9 +15,100 @@
...
+
+      // Iterate through attendees to check the events each of them are attending
+      for (Iterator<String> j = attendees.iterator(); j.hasNext();) {
+        String attendee = j.next();
+        if (event.getAttendees().contains(attendee)) {
+
+          // Iterate through all time periods to see
+          // if any period overlaps with an event the attendees have
+          for (ListIterator<TimeRange> k = timeSlots.listIterator(); k.hasNext();) {
+            TimeRange currTime = k.next();
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Perhaps rename this variable for clarity? My initial expectation was that this refers to the current wall clock time.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -15,9 +15,100 @@
...
+            // Split time period from B into 1 and 2
+            } else if (currTime.contains(eventTime)) {
+              k.remove();
+              k.add(TimeRange.fromStartEnd(currTime.start(), eventTime.start(), false));
+              k.add(TimeRange.fromStartEnd(eventTime.end(), currTime.end(), false));
+            
+            // Case C: |--C---|
+            // Event:      |------|
+            //         |-1-|
+            // Shorten time period from C to 1
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: 1) Good use of ASCII art to clarify the algorithm's different scenarios, I like it!
2) Do 'slot' and 'period' mean the same thing? If yes, I'd suggest standardizing for clarity.
3) I like how one timeSlot is named "Event". To be complete, what would be the name of the other two timeslots (lines 83 and 85)?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -15,9 +15,100 @@
...
+              }
+            }
+          }
+        }
+      }
+    }
+    return timeSlots;
+  }
+  
+  // Delete any time slots that has a duration less than the request time
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: typo: has --> have

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 14]
```diff
@@ -270,5 +272,149 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
+
+  @Test
+  public void optionalAttendeeNotConsidered() {
+    // Have each person have different events. We should see three options because each person has
+    // split the restricted times.
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: What does "each person has split the restricted times" mean?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 15]
```diff
@@ -270,5 +272,149 @@ public void notEnoughRoom() {
...
+        Arrays.asList(TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void noRequiredAttendee() {
+    // Have each person have different events. We should see three options because each person has
+    // split the restricted times.
+    // Optional attendee has all day event. We should still return the three options
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Is it correct that one of the optional attendees A or B has an all day event? Doesn't match the ASCII art below.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 16]
```diff
@@ -270,5 +272,149 @@ public void notEnoughRoom() {
...
+    MeetingRequest request =
+        new MeetingRequest(NO_ATTENDEES, DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected = Arrays.asList();
+
+    Assert.assertEquals(expected, actual);
+  }
```
Comment (mutisog) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:422]: - Should we add a test for the case where no time slots are available?
- What other edge cases should be considered / are worth testing?

**Review by mutisog — COMMENTED** [review]
Left some comments.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (yjzhang111) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:46]: I am saving a copy of timeSlots to variable timeSlotsForRequired (in the case where there is no time slot available for optional attendees, I can just return timeSlotsForRequired).

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (yjzhang111) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: The variable timeSlots is first modified to fit all the required attendees' schedules, and the result is saved in timeSlotsForRequired. Then it goes on to check all the optional attendees, so after line 43, timeSlots include slots for both required and optional attendees. Do you think it is better to change the variable name to something like timeSlotsForAll? (I think the only problem with this is that making a copy of timeSlotsForAll and naming it timeSlotsForRequired does not make a lot of sense) Or do you think it is better to just add some comments here?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 16 above)
Comment (yjzhang111) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:422]: Some other tests are provided in the file as starter code (when the attendees have overlapping/nested events, when there is no time slots available, when there is no attendee at all, etc.)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 14 above)
Comment (yjzhang111) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Basically means the events each attendee has split the whole day into 3 options for the request. I've changed the wording so that it's less confusing.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (yjzhang111) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: 2) Fixed. 3) I changed Case C and event to their corresponding variable name so it matches with the code better.
